src/js/components/DataTable/DataTable.js

```diff
@@ -314,6 +320,80 @@ const DataTable = ({
       ? { style: { minWidth: '100%' } }
       : undefined;
 
+  // Choose what to display: button to load more , spinner, or
+  // unretrieved message in case the function takes long to get data
+  const [moreButtonState, setMoreButtonState] = useState('button');
+
+  const moreButtonRef = useRef(moreButtonState);
+  moreButtonRef.current = moreButtonState;
+
+  // If 10 seconds pass without getting data change moreButtonState
+  // to let user know and ask to load more again
+
+  const afterTenSecondsRef = useRef(null);
+  const dataTimeout = () => {
+    afterTenSecondsRef.current = setTimeout(() => {
+      if (moreButtonRef.current === 'spinner') {
+        setMoreButtonState('unretrieved');
+      }
+    }, 10000);
+  };
+
+  useEffect(() => {
+    if (moreButtonState === 'spinner') {
+      onMore();
+      dataTimeout();
+    }
+  }, [moreButtonState]);
+
+  // When data is retrieved set moreButton again
+  useEffect(() => {
+    setMoreButtonState('button');
+    if (afterTenSecondsRef.current) {
+      clearTimeout(afterTenSecondsRef.current);
+      afterTenSecondsRef.current = null;
+    }
+  }, [data.length]);
+
+  if (moreButton === true && !onMore) {
+    console.warn(
+      'You need to provide an function in the onMore field to use moreButton',
+    );
+  }
+
+  const memoizedmoreButtonState = useMemo(() => {
+    if (moreButton && moreButtonState === 'button') {
```

Comment:
Feels like having all of these button states handled as strings introduces opportunity for things to break. What about creating some constant and referencing here?

```
const moreButtonStates = {
   button: 'button',
   spinner: 'spinner',
   unretrieved: 'unretrieved'
}

...

moreButton && moreButtonState === moreButtonStates.button
```